include/dali/core/mm/detail/free_list.h

```diff
@@ -645,6 +645,101 @@ class coalescing_free_tree {
   detail::pooled_set<std::pair<size_t, char *>, true> by_size_;
 };
 
+/**
+ * @brief Maintains a tree of free memory blocks of variable size, returning free blocks
+ *        with least margin.
+ *
+ * This free tree does not combine blocks - it returns the one with least margin, if
+ * the margin is within the limit. Maximum size of a suitable block is relative
+ * to the requested block.
+ *
+ * When requesting a block smaller than the one in the tree, the information about the original
+ * block is stored in a special structure and restored upon deallocation, so the entire original
+ * block can be reconstituted.
+ */
+class best_fit_free_tree {
+ public:
+  void clear() {
+    by_addr_.clear();
+    by_size_.clear();
+    original_.clear();
+  }
+
+  float max_padding_ratio = 1.1f;
+
+  /**
+   * @brief Gets a block that has at least the specified size and alignment
+   *
+   * This free tree will not use blocks larger than `size * max_padding_ratio`
+   */
+  void *get(size_t size, size_t alignment) {
+    // the formula below is to avoid rounding errors - do not "optimize"
+    size_t max_size = size + static_cast<size_t>(size * (max_padding_ratio - 1));
+
+    for (auto it = by_size_.lower_bound({ size, nullptr }); it != by_size_.end(); ++it) {
+      size_t block_size = it->first;
+      if (block_size > max_size)
+        break;  // nothing good will happen
+      char *base = it->second;
+      char *aligned = detail::align_ptr(base, alignment);
+      by_size_.erase(it);
+      by_addr_.erase(base);
+      if (block_size != size)  // only store if there's padding
```

Comment:
Maybe for simplicity you can remove that if.
Just a thought.

Reply:
No, it's not good. If I remove a block completely, I can put a block of different size in the list and having it in original_ would interfere.